File: src/#430/solution.go
package solution

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	if root == nil {
		return nil
	}

	stack := []*Node{}
	current := root
	for current.Child != nil || current.Next != nil {
		if current.Child != nil {
			if current.Next != nil {
				stack = append(stack, current.Next)
			}

			prev := current
			current.Next = current.Child
			current.Child = nil
			current = current.Next
			current.Prev = prev
		} else {
			current = current.Next
		}
	}

	prev := current
	for len(stack) != 0 {
		if current.Next == nil {
			popupNode := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			current.Next = popupNode
			current = current.Next
			current.Prev = prev
			prev = current
		} else {
			current = current.Next
			prev = current
		}
	}

	return root
}

File: src/#430/solution.js
function Node(val, prev, next, child) {
  this.val = val
  this.prev = prev
  this.next = next
  this.child = child
}

var flatten = function (head) {
  if (head === null) {
    return null
  }

  let stack = []
  let current = head

  while (current.child !== null || current.next !== null) {
    if (current.child !== null) {
      if (current.next !== null) {
        stack.push(current.next)
      }

      let prev = current
      current.next = current.child
      current.child = null
      current = current.next
      current.prev = prev

    } else {
      current = current.next
    }
  }

  let prev = current

  while (stack.length !== 0) {
    if (current.next === null) {

      let popupNode = stack.pop()
      current.next = popupNode
      current = current.next
      current.prev = prev
      prev = current

    } else {

      current = current.next
      current.prev = prev
      prev = current
      
    }
  }
  return head
}
